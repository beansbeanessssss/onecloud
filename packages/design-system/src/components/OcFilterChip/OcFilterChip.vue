<template>
  <div class="oc-filter-chip oc-flex" :class="{ 'oc-filter-chip-toggle': isToggle }">
    <oc-button
      :id="id"
      class="oc-filter-chip-button oc-pill"
      :class="{ 'oc-filter-chip-button-selected': filterActive }"
      appearance="raw"
      @click="isToggle ? $emit('toggleFilter') : false"
    >
      <oc-icon v-if="filterActive" name="check" size="small" color="var(--oc-color-text-inverse)" />
      <span
        class="oc-text-truncate oc-filter-chip-label"
        v-text="!!selectedItemNames.length ? selectedItemNames[0] : filterLabel"
      />
      <span v-if="selectedItemNames.length > 1" v-text="` +${selectedItemNames.length - 1}`" />
      <oc-icon v-if="!filterActive && !isToggle" name="arrow-down-s" size="small" />
    </oc-button>
    <oc-drop
      v-if="!isToggle"
      :toggle="'#' + id"
      class="oc-filter-chip-drop"
      mode="click"
      padding-size="small"
      @hide-drop="$emit('hideDrop')"
      @show-drop="$emit('showDrop')"
    >
      <slot />
    </oc-drop>
    <oc-button
      v-if="filterActive"
      v-oc-tooltip="$gettext('Clear filter')"
      class="oc-filter-chip-clear oc-px-xs"
      appearance="raw"
      @click="$emit('clearFilter')"
    >
      <oc-icon name="close" size="small" color="var(--oc-color-text-inverse)" />
    </oc-button>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue'
import uniqueId from '../../utils/uniqueId'

export default defineComponent({
  name: 'OcFilterChip',
  status: 'ready',
  release: '15.0.0',
  props: {
    id: {
      type: String,
      required: false,
      default: () => uniqueId('oc-filter-chip-')
    },
    /**
     * Label which is displayed when no items are selected.
     */
    filterLabel: {
      type: String,
      required: true
    },
    /**
     * An array of selected item names. It is being ignored when `isToggle` is set to `true`.
     */
    selectedItemNames: {
      type: Array,
      required: false,
      default: () => []
    },
    /**
     * Display the filter chip as a on/off toggle.
     */
    isToggle: {
      type: Boolean,
      default: false
    },
    /**
     * Whether the toggle filter is active. It only has an effect if `isToggle` is set to `true`.
     */
    isToggleActive: {
      type: Boolean,
      default: false
    }
  },
  emits: ['clearFilter', 'hideDrop', 'showDrop', 'toggleFilter'],
  setup(props) {
    const filterActive = computed(() => {
      if (props.isToggle) {
        return props.isToggleActive
      }
      return !!props.selectedItemNames.length
    })
    return { filterActive }
  }
})
</script>

<style lang="scss">
.oc-filter-chip {
  &-button {
    align-items: center;
    background-color: var(--oc-color-background-default) !important;
    color: var(--oc-color-text-muted) !important;
    border: 1px solid var(--oc-color-text-muted);
    box-sizing: border-box;
    display: inline-flex;
    gap: var(--oc-space-xsmall);
    text-decoration: none;
    font-size: var(--oc-font-size-xsmall);
    line-height: 1rem;
    max-width: 150px;
    height: 24px;
    padding: var(--oc-space-xsmall) var(--oc-space-small) !important;

    &-selected,
    &-selected:hover {
      background-color: var(--oc-color-swatch-primary-default) !important;
      color: var(--oc-color-text-inverse) !important;
      border-top-left-radius: 99px !important;
      border-bottom-left-radius: 99px !important;
      border-top-right-radius: 0px !important;
      border-bottom-right-radius: 0px !important;
      border: 0;
    }
  }
  &-clear,
  &-clear:hover {
    background-color: var(--oc-color-swatch-primary-default) !important;
    color: var(--oc-color-text-inverse) !important;
    border-top-left-radius: 0px !important;
    border-bottom-left-radius: 0px !important;
    border-top-right-radius: 99px !important;
    border-bottom-right-radius: 99px !important;
  }
  &-clear:not(.oc-filter-chip-toggle .oc-filter-chip-clear),
  &-clear:hover:not(.oc-filter-chip-toggle .oc-filter-chip-clear) {
    margin-left: 1px;
  }
}
</style>
