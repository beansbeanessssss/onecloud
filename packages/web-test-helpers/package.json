{
  "name": "web-test-helpers",
  "private": true,
  "main": "src/index.ts",
  "peerDependencies": {
    "axios": "0.27.2",
    "vue-gettext": "2.1.12",
    "vue-router": "3.6.5",
    "vuex": "4.1.0"
  }
}
