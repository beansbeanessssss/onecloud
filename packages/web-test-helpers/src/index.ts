export * from './mocks/defaultComponentMocks'
export * from './mocks/store/defaultStoreMockOptions'
export * from './mocks/defaultStubs'
export * from './mocks/axios'
export * from './defaultPlugins'
export * from './helpers'
