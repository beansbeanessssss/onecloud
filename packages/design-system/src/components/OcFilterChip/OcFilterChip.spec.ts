import { defaultPlugins, mount } from 'web-test-helpers'
import OcFilterChip from './OcFilterChip.vue'

const selectors = {
  clearBtn: '.oc-filter-chip-clear'
}

describe('OcFilterChip', () => {
  it('renders the filterName when no selected item has been given', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('renders the first selected item name when selected items have been given', () => {
    const { wrapper } = getWrapper({ props: { selectedItemNames: ['Einstein', 'Marie'] } })
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('emits the "clearFilter"-event when clicking the clear-button', async () => {
    const { wrapper } = getWrapper({ props: { selectedItemNames: ['Einstein', 'Marie'] } })
    await wrapper.find(selectors.clearBtn).trigger('click')
    expect(wrapper.emitted('clearFilter')).toBeTruthy()
  })
})

const getWrapper = ({ props = {} } = {}) => {
  return {
    wrapper: mount(OcFilterChip, {
      props: { filterName: 'Users', selectedItemNames: [], ...props },
      global: { plugins: [...defaultPlugins()] }
    })
  }
}
