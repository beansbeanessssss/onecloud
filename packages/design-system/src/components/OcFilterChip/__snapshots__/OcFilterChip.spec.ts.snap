// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OcFilterChip renders the filterName when no selected item has been given 1`] = `
<div class="oc-filter-chip oc-flex">
  <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-filter-chip-button" id="oc-filter-chip-1" type="button">
    <!-- @slot Content of the button -->
    <!--v-if-->
    <span class="oc-text-truncate">Users</span>
    <!--v-if-->
    <span class="oc-icon oc-icon-s oc-icon-passive">
      <!---->
    </span>
  </button>
  <div class="oc-drop oc-box-shadow-medium oc-rounded oc-filter-chip-drop" id="oc-drop-2">
    <div class="oc-card oc-card-body oc-background-secondary oc-p-m"></div>
  </div>
  <!--v-if-->
</div>
`;

exports[`OcFilterChip renders the first selected item name when selected items have been given 1`] = `
<div class="oc-filter-chip oc-flex">
  <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-filter-chip-button oc-filter-chip-button-selected" id="oc-filter-chip-3" type="button">
    <!-- @slot Content of the button -->
    <span class="oc-icon oc-icon-s oc-icon-passive">
      <!---->
    </span>
    <span class="oc-text-truncate">Einstein</span>
    <span> +1</span>
    <!--v-if-->
  </button>
  <div class="oc-drop oc-box-shadow-medium oc-rounded oc-filter-chip-drop" id="oc-drop-4">
    <div class="oc-card oc-card-body oc-background-secondary oc-p-m"></div>
  </div>
  <button class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw oc-filter-chip-clear oc-px-xs" type="button">
    <!-- @slot Content of the button -->
    <span class="oc-icon oc-icon-s oc-icon-passive">
      <!---->
    </span>
  </button>
</div>
`;
